plugins/network-elements/nuage-vsp/src/com/cloud/network/guru/NuageVspGuestNetworkGuru.java

```diff
@@ -201,51 +234,82 @@ public NicProfile allocate(Network network, NicProfile nic, VirtualMachineProfil
+            boolean isL3Network = isL3Network(network);
+            boolean isSharedNetwork = networkOffering.getGuestType() == GuestType.Shared;
+            Long vpcId = network.getVpcId();
+            String vpcUuid = null;
+            if (vpcId != null) {
+                Vpc vpcObj = _vpcDao.findById(vpcId);
+                vpcUuid = vpcObj.getUuid();
+            }
+            HostVO nuageVspHost = getNuageVspHost(network.getPhysicalNetworkId());
+            IPAddressVO staticNatIp = _ipAddressDao.findByVmIdAndNetworkId(network.getId(), vm.getId());
+
+            ReserveVmInterfaceVspCommand cmd;
+            if (staticNatIp != null) {
+                VlanVO staticNatVlan = _vlanDao.findById(staticNatIp.getVlanId());
+                cmd = new ReserveVmInterfaceVspCommand(nicFrmDB.getUuid(), nic.getMacAddress(), network.getUuid(), isL3Network, isSharedNetwork, vpcUuid,
+                        networksDomain.getUuid(), networksAccount.getUuid(), isDomainRouter, domainRouterIp,
+                        vm.getInstanceName(), vm.getUuid(), networksDomain.getUuid(), networksAccount.getUuid(), true, nic.getIPv4Address(), staticNatIp.getUuid(),
+                        staticNatIp.getAddress().addr(), staticNatIp.getState().equals(IpAddress.State.Allocated), staticNatIp.isOneToOneNat(), staticNatVlan.getUuid(),
+                        staticNatVlan.getVlanGateway(), staticNatVlan.getVlanNetmask());
+            } else {
+                cmd = new ReserveVmInterfaceVspCommand(nicFrmDB.getUuid(), nic.getMacAddress(), network.getUuid(), isL3Network, isSharedNetwork, vpcUuid,
+                        networksDomain.getUuid(), networksAccount.getUuid(), vm.getType().equals(VirtualMachine.Type.DomainRouter), domainRouterIp,
+                        vm.getInstanceName(), vm.getUuid(), networksDomain.getUuid(), networksAccount.getUuid(), true, nic.getIPv4Address(), null, null, false, false,
+                        null, null, null);
+            }
+
+            ReserveVmInterfaceVspAnswer answer = (ReserveVmInterfaceVspAnswer)_agentMgr.easySend(nuageVspHost.getId(), cmd);
+            if (answer == null || !answer.getResult()) {
+                s_logger.error("ReserveVmInterfaceNuageVspCommand failed");
+                if ((null != answer) && (null != answer.getDetails())) {
+                    s_logger.error(answer.getDetails());
+                }
+                throw new InsufficientVirtualNetworkCapacityException("Failed to reserve VM in Nuage VSP.", Network.class, network.getId());
+            }
+
+            if (isDomainRouter) {
+                nic.setIPv4Address(domainRouterIp);
+            }
+
+        } finally {
+            if (network != null && lockedNetwork) {
+                _networkDao.releaseFromLockTable(network.getId());
+                s_logger.debug("Unlocked network " + network.getId() + " for creation of user VM " + vm.getInstanceName());
```

Comment:
Wrap in a if (s_logger.isDebugEnabled()) { ... } block to avoid unnecessary pressure on the string pool when debug logging is not enabled.